go/database/mpt/proof/verification.go

```diff
@@ -0,0 +1,465 @@
+	trie, err := mpt.OpenArchiveTrie(dir, config, mpt.NodeCacheConfig{}, mpt.ArchiveConfig{})
+	if err != nil {
+		return err
+	}
+
+	observer.StartVerification()
+	err = errors.Join(
+		verifyArchiveTrie(ctx, trie, from, to, observer),
+		trie.Close(),
+	)
+	observer.EndVerification(err)
+	return err
+}
+
+func verifyArchiveTrie(ctx context.Context, trie verifiableArchiveTrie, from, to int, observer mpt.VerificationObserver) error {
+	blockHeight, empty, err := trie.GetBlockHeight()
+	if err != nil {
+		return err
+	}
+	if empty {
+		return nil
+	}
+
+	if to > int(blockHeight) {
+		to = int(blockHeight)
+		observer.Progress(fmt.Sprintf("setting a maximum block height: %d", blockHeight))
+	}
+
+	observer.Progress(fmt.Sprintf("Verifying total block range [%d;%d]", from, to))
+	for i := from; i <= to; i++ {
+		trieView := &archiveTrie{trie: trie, block: uint64(i)}
+		observer.Progress(fmt.Sprintf("Verifying block: %d ", i))
+		if err := verifyTrie(ctx, trieView, observer); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// VerifyLiveTrie verifies the consistency of witness proofs for a live trie.
+// It reads the trie for the head block and loads accounts and storage slots.
+// It extracts witness proofs for these accounts and its storage,
+// and checks that values in the proof and the database match.
+func VerifyLiveTrie(ctx context.Context, dir string, config mpt.MptConfig, observer mpt.VerificationObserver) error {
+	trie, err := mpt.OpenFileLiveTrie(dir, config, mpt.NodeCacheConfig{})
+	if err != nil {
+		return err
+	}
+
+	observer.StartVerification()
+	err = errors.Join(
+		verifyTrie(ctx, trie, observer),
+		trie.Close(),
+	)
+	observer.EndVerification(err)
+	return err
+}
+
+func verifyTrie(ctx context.Context, trie verifiableTrie, observer mpt.VerificationObserver) error {
+	rootHash, hints, err := trie.UpdateHashes()
+	if hints != nil {
+		hints.Release()
+	}
+	if err != nil {
+		return err
+	}
+
+	observer.Progress("Collecting and Verifying proofs ... ")
+	visitor := accountVerifyingVisitor{
+		ctx:       ctx,
+		rootHash:  rootHash,
+		trie:      trie,
+		observer:  observer,
+		logWindow: 1000_000,
+	}
+	if err := trie.VisitTrie(&visitor); err != nil || visitor.err != nil {
+		return errors.Join(err, visitor.err)
+	}
+
+	const numAddresses = 1000
+	return verifyEmptyAccount(trie, rootHash, numAddresses, observer)
+}
+
+// verifyEmptyAccount verifies the consistency of witness proofs for empty accounts that are not present in the trie.
+func verifyEmptyAccount(trie verifiableTrie, rootHash common.Hash, numAddresses int, observer mpt.VerificationObserver) error {
+	observer.Progress(fmt.Sprintf("Veryfing %d empty addresses...", numAddresses))
+	addresses, err := generateUnusedAddresses(trie, numAddresses)
+	if err != nil {
+		return err
+	}
+	for _, addr := range addresses {
+		proof, err := trie.CreateWitnessProof(addr)
```

Comment:
The `proof` produced here seems to be never checked for being consistent using its [`IsValid()`](https://github.com/Fantom-foundation/Carmen/blob/3fc9e177f6bb15f517db30573f9b50266706de04/go/common/witness/proof.go#L36) member function. Since the aim of the proof tool is mainly to make sure that the extracted proofs are correct, I think this should be added -- likely also in other places. Right?

Reply:
ok, added